chain/ethereum/ethereum.go

```diff
@@ -183,6 +186,66 @@ func ChunkedFilterLogs(ctx context.Context, client ETHClient, addresses []common
 	return failedQueries, nil
 }
 
+// BatchedFilterLogs queries the blockchain for past events in batches.
+// It slices addresses and total number of blocks with pre-defined batch size.
+// The results are later fed into a log chan that was provided by the caller.
+// Failed query intervals are retried after a backoff period, increased exponentially.
+func BatchedFilterLogs(ctx context.Context, client ETHClient, addresses []common.Address, fromBlock, toBlock uint64, logChan chan<- types.Log, backoff time.Duration) {
+	if backoff == 0 || backoff > maxBackoffDuration {
+		backoff = initialBackoffDuration
+	}
+
+	//	Slice and dice whole block interval and smart contract address count so that they fit into predefined maximums.
+	if len(addresses) > addressChunkSize {
```

Comment:
I think it is better to change the order of the code from this:
```
	if len(addresses) > addressChunkSize {
		BatchedFilterLogs(ctx, client, addresses[:addressChunkSize], fromBlock, toBlock, logChan, backoff)
		BatchedFilterLogs(ctx, client, addresses[addressChunkSize:], fromBlock, toBlock, logChan, backoff)
		return
	}

	if toBlock-fromBlock > blockChunkSize {
		BatchedFilterLogs(ctx, client, addresses, toBlock-blockChunkSize, toBlock, logChan, backoff)
		BatchedFilterLogs(ctx, client, addresses, fromBlock, toBlock-blockChunkSize-1, logChan, backoff)
		return
	}
```
to this:
```
	if toBlock-fromBlock > blockChunkSize {
		BatchedFilterLogs(ctx, client, addresses, toBlock-blockChunkSize, toBlock, logChan, backoff)
		BatchedFilterLogs(ctx, client, addresses, fromBlock, toBlock-blockChunkSize-1, logChan, backoff)
		return
	}

	if len(addresses) > addressChunkSize {
		BatchedFilterLogs(ctx, client, addresses[:addressChunkSize], fromBlock, toBlock, logChan, backoff)
		BatchedFilterLogs(ctx, client, addresses[addressChunkSize:], fromBlock, toBlock, logChan, backoff)
		return
	}
```
so that it will backfill all contracts concurrently or else it will wait for first `addressChunkSize` contracts to finish backfilling before starting backfilling next contracts in the list.